File: examples/ruby/error_handling.rb
# frozen_string_literal: true

# Rescue the typed errors raised by the Pushfy Ruby SDK.
#
#   PUSHFY_API_TOKEN=... ruby error_handling.rb
#
# Every failure raises a subclass of Pushfy::PushfyError, each exposing
# #status (HTTP status, 0 for network/timeout), #code (API error string)
# and #response (parsed body).

require "pushfy"

def main
  pushfy = Pushfy::Client.new(api_token: ENV.fetch("PUSHFY_API_TOKEN"))

  pushfy.sms.send(to: "5511999999999", text: "Hello from Pushfy")
  puts "Sent."
rescue KeyError => e
  warn "Missing environment variable: #{e.message}"
  exit 1
rescue Pushfy::AuthenticationError => e
  # 401/403 — token missing/invalid or bad HMAC signature. Do NOT retry.
  warn "Authentication failed (#{e.status}): check your API token."
  exit 1
rescue Pushfy::InvalidRequestError => e
  # 4xx — malformed request. Fix the payload; retrying won't help.
  warn "Invalid request (#{e.status} #{e.code}): #{e.response.inspect}"
  exit 1
rescue Pushfy::RateLimitError => e
  # 429 — slow down. retry_after is seconds when the API provides it.
  wait = e.retry_after || 5
  warn "Rate limited — retry after #{wait}s."
  exit 1
rescue Pushfy::ApiError => e
  # 5xx / network / timeout — safe to retry idempotently (reuse the ext_id).
  warn "Transient API error (#{e.status}): #{e.message}"
  exit 1
rescue Pushfy::PushfyError => e
  # Catch-all for anything new the SDK might add.
  warn "Unexpected Pushfy error: #{e.status} #{e.code} #{e.response.inspect}"
  exit 1
end

main

File: examples/typescript/error-handling.ts
/**
 * error-handling.ts — Branch on typed Pushfy errors with `instanceof`.
 *
 * Every failure throws a subclass of PushfyError, each carrying
 * `status`, `code` and `response`. Handle the specific ones first,
 * then fall back to the generic ApiError (5xx / network).
 *
 * Run:  PUSHFY_API_TOKEN=... npx ts-node error-handling.ts
 */
import {
  Pushfy,
  PushfyError,
  AuthenticationError,
  RateLimitError,
  InvalidRequestError,
  ApiError,
} from '@pushfy/pushfy';

async function main(): Promise<void> {
  const pushfy = new Pushfy({ apiToken: process.env.PUSHFY_API_TOKEN ?? 'invalid-token' });

  try {
    await pushfy.sms.send({
      to: '5511999999999',
      text: 'Hello from Pushfy',
      extId: 'err-demo-001',
    });
    console.log('Sent successfully.');
  } catch (err: unknown) {
    if (err instanceof RateLimitError) {
      // 429 — slow down. err.retryAfter (seconds) may be present.
      console.error(`Rate limited. Retry after ${err.retryAfter ?? '?'}s.`);
    } else if (err instanceof AuthenticationError) {
      // 401/403 — bad token or HMAC keys.
      console.error('Authentication failed — check your credentials.');
    } else if (err instanceof InvalidRequestError) {
      // 400/413/415 — the request itself is wrong; do NOT retry as-is.
      console.error(`Invalid request (${err.status}): ${err.message}`);
    } else if (err instanceof ApiError) {
      // 5xx or network. Safe to retry idempotently (reuse the same extId).
      console.error(`Server/network error (${err.status}). Retry idempotently.`);
    } else if (err instanceof PushfyError) {
      // Any other Pushfy error.
      console.error(`Pushfy error (${err.status}/${err.code}): ${err.message}`);
    } else {
      // Not a Pushfy error at all.
      throw err;
    }

    if (err instanceof PushfyError) {
      console.error('Raw API response:', err.response);
    }
    process.exitCode = 1;
  }
}

main().catch((err: unknown) => {
  console.error('Unhandled error:', err instanceof Error ? err.message : err);
  process.exit(1);
});
